python/openai/openai/src/routers/chat_completions.py

```diff
@@ -0,0 +1,200 @@
+    CreateChatCompletionRequest,
+    CreateChatCompletionResponse,
+    CreateChatCompletionStreamResponse,
+    ObjectType,
+)
+from src.utils.triton import get_output, validate_triton_responses
+
+router = APIRouter()
+
+
+def get_first_response_role(conversation, add_generation_prompt, default_role):
+    if add_generation_prompt:
+        return default_role
+
+    return conversation[-1]["role"]
+
+
+def streaming_chat_completion_response(request_id, created, model, role, responses):
+    # first chunk
+    choice = ChatCompletionStreamingResponseChoice(
+        index=0,
+        delta=ChatCompletionStreamResponseDelta(
+            role=role, content="", function_call=None
+        ),
+        logprobs=None,
+        finish_reason=None,
+    )
+    chunk = CreateChatCompletionStreamResponse(
+        id=request_id,
+        choices=[choice],
+        created=created,
+        model=model,
+        system_fingerprint=None,
+        object=ObjectType.chat_completion_chunk,
+    )
+    yield f"data: {chunk.json(exclude_unset=True)}\n\n"
+
+    for response in responses:
+        text = get_output(response)
+
+        choice = ChatCompletionStreamingResponseChoice(
+            index=0,
+            delta=ChatCompletionStreamResponseDelta(
+                role=None, content=text, function_call=None
+            ),
+            logprobs=None,
+            finish_reason=ChatCompletionFinishReason.stop if response.final else None,
+        )
+
+        chunk = CreateChatCompletionStreamResponse(
+            id=request_id,
+            choices=[choice],
+            created=created,
+            model=model,
+            system_fingerprint=None,
+            object=ObjectType.chat_completion_chunk,
+        )
+
+        yield f"data: {chunk.json(exclude_unset=True)}\n\n"
+
+    yield "data: [DONE]\n\n"
+
+
+@router.post(
+    "/v1/chat/completions", response_model=CreateChatCompletionResponse, tags=["Chat"]
+)
+def create_chat_completion(
+    request: CreateChatCompletionRequest,
+    raw_request: Request,
+) -> CreateChatCompletionResponse | StreamingResponse:
+    """
+    Creates a model response for the given chat conversation.
+    """
+
+    model_metadatas = raw_request.app.models
+    if not model_metadatas:
+        raise HTTPException(status_code=400, detail="No known models")
+
+    metadata = model_metadatas.get(request.model)
+    if not metadata:
+        raise HTTPException(status_code=400, detail=f"Unknown model: {request.model}")
+
+    if not metadata.request_convert_fn:
+        raise HTTPException(
+            status_code=400, detail=f"Unknown request format for model: {request.model}"
+        )
+
+    if not metadata.tokenizer:
+        raise HTTPException(status_code=400, detail="Unknown tokenizer")
+
+    if not metadata.backend:
+        raise HTTPException(status_code=400, detail="Unknown backend")
+
+    triton_model = raw_request.app.server.model(request.model)
+    if request.model != triton_model.name:
```

Comment:
curious: I'm not sure how this would be possible - the model api doesn't do any checks and would return exactly what you give it ... was there an issue that you saw?

Reply:
This looks to be redundant, probably from some find/replace, will remove :+1: